modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryUtils.cs

```diff
@@ -1084,7 +1084,7 @@ public static void WriteArray(Array val, BinaryWriter ctx, int? elemTypeId = nul
             stream.WriteInt(val.Length);
 
             for (int i = 0; i < val.Length; i++)
-                ctx.Write(val.GetValue(i));
+                ctx.WriteObjectDetached(val.GetValue(i), parentCollection: val);
```

Comment:
Don't you think that using `Write` instead of `WriteObjectDetached` is very error-prone and sooner or later produce a hard-to-find errors? Are there cases where `Write` really gives a good performance gain?

Reply:
`Write` supports circular references, so we can't write everything in detached mode.

Yes, this feature introduces a lot of complexity, and possibly degrades performance due to handle tracking. Hopefully we'll get rid of this on 3.0, but for 2.x we have to maintain the compatibility.